* Experiment on the dataset *sc_10x_5cl*

** Data preparation
- *bulk_RNAseq* directory contains the bulkRNAseq data (GSE86337) directly from
  the NCBI GEO website.
- get_DEgenes_from_bulkRNAseq.R is used to generate the differential expressed
  genes from the bulkRNAseq data based on the tool DESeq2.
- prepare_scRNAseq.R is used to generate the scRNAseq based on the GitHub repository:
  sc_mixology. In the code, we describe how to get the data we use. Part of the
  codes are learned from the Gihub repository: imputationBenchmark.

** How to run
- Our jobs are organized and executed by *make* under the *jobs* directory.
- Our jobs are submitted into the *slurm* to run. But you can get it run locally
  by following the shell scripts under *jobs*. 
- impute_cluster_de.R is used for imputation, clustering and differential
  expressed gene analysis.

** Data Summary
- result_summary.R is used to generate Supplementary Text: Figure S17-18.
